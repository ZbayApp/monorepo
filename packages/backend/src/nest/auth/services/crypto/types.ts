import { Base58, SignedEnvelope } from '@localfirst/auth'

export enum EncryptionScopeType {
  ROLE = 'ROLE',
  CHANNEL = 'CHANNEL',
  USER = 'USER',
  TEAM = 'TEAM',
}

export type EncryptionScope = {
  type: EncryptionScopeType
  name?: string
}

export type EncryptedPayload = {
  contents: Base58
  scope: EncryptionScope & {
    generation: number
  }
}

export type EncryptedAndSignedPayload = {
  encrypted: EncryptedPayload
  signature: SignedEnvelope
  ts: number
  username: string
}
