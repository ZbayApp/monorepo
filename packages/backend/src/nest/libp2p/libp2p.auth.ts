import {
  ComponentLogger,
  Connection,
  NewStreamOptions,
  PeerId,
  PeerStore,
  serviceCapabilities,
  serviceDependencies,
  Stream,
  Topology,
} from '@libp2p/interface'
import type { ConnectionManager, IncomingStreamData, Registrar } from '@libp2p/interface-internal'
import * as Auth from '@localfirst/auth'
import { pushable, type Pushable } from 'it-pushable'
import { Uint8ArrayList } from 'uint8arraylist'
import { pipe } from 'it-pipe'
import { encode, decode } from 'it-length-prefixed'

import { SigChainService } from '../auth/sigchain.service'
import { createLogger } from '../common/logger'
import { createQuietLogger } from '@quiet/logger'
import { ConnectionParams } from '3rd-party/auth/packages/auth/dist/connection/Connection'
import { Libp2pService } from './libp2p.service'
import { Libp2pEvents } from './libp2p.types'

export interface Libp2pAuthComponents {
  peerId: PeerId
  peerStore: PeerStore
  registrar: Registrar
  connectionManager: ConnectionManager
  logger: ComponentLogger
}

export interface Libp2pAuthStatus {
  started: boolean
  joining: boolean
}

export interface PushableStream {
  stream: Stream
  pushable: Pushable<Uint8Array | Uint8ArrayList>
}

enum JoinStatus {
  PENDING = 'PENDING',
  JOINING = 'JOINING',
  JOINED = 'JOINED',
}

const createLFALogger = createQuietLogger('localfirst')

export class Libp2pAuth {
  private readonly protocol: string
  private readonly components: Libp2pAuthComponents
  private sigChainService: SigChainService
  private libp2pService: Libp2pService
  private authConnections: Record<string, Auth.Connection>
  private outboundStreamQueue: Pushable<{ peerId: PeerId; connection: Connection }>
  private outboundStreams: Record<string, PushableStream>
  private inboundStreams: Record<string, Stream>
  private bufferedConnections: { peerId: PeerId; connection: Connection }[]
  private joining: boolean = false
  private unblockInterval: NodeJS.Timeout
  private joinStatus: JoinStatus
  private logger: ReturnType<typeof createLogger> = createLogger('libp2p:auth')
  readonly [serviceCapabilities]: string[] = ['@quiet/auth']
  readonly [serviceDependencies]: string[] = ['@libp2p/identify']
  readonly [Symbol.toStringTag]: string = 'lfaAuth'

  constructor(sigChainService: SigChainService, libp2pService: Libp2pService, components: Libp2pAuthComponents) {
    this.protocol = '/local-first-auth/1.0.0'
    this.components = components
    this.sigChainService = sigChainService
    this.libp2pService = libp2pService
    this.authConnections = {}
    this.outboundStreamQueue = pushable<{ peerId: PeerId; connection: Connection }>({ objectMode: true })
    this.outboundStreams = {}
    this.inboundStreams = {}
    this.bufferedConnections = []
    this.joinStatus = JoinStatus.PENDING
    this.logger = this.logger.extend(sigChainService.getActiveChain().localUserContext.user.userName)

    this.logger.info('Auth service initialized')
    this.logger.info('sigChainService', sigChainService.activeChainTeamName)

    pipe(this.outboundStreamQueue, async source => {
      for await (const { peerId, connection } of source) {
        await this.openOutboundStream(peerId, connection)
      }
    }).catch(e => {
      this.logger.error('Outbound stream queue error', e)
    })

    this.unblockInterval = setInterval(this.unblockConnections, 5_000, this.bufferedConnections, this.joinStatus)
  }

  private emit(eventName: string, ...args: any[]) {
    this.libp2pService.emit(eventName, ...args)
  }

  private async unblockConnections(conns: { peerId: PeerId; connection: Connection }[], status: JoinStatus) {
    if (status !== JoinStatus.JOINED) return

    this.logger.info(`Unblocking ${conns.length} connections now that we've joined the chain`)
    while (conns.length > 0) {
      const conn = conns.pop()
      if (conn != null) {
        await this.onPeerConnected(conn.peerId, conn.connection)
      }
    }
  }

  async start() {
    this.logger.info('Auth service starting')

    const topology: Topology = {
      onConnect: this.onPeerConnected.bind(this),
      onDisconnect: this.onPeerDisconnected.bind(this),
      notifyOnLimitedConnection: false,
    }

    const registrar = this.components.registrar
    await registrar.register(this.protocol, topology)
    await registrar.handle(this.protocol, this.onIncomingStream.bind(this), {
      runOnLimitedConnection: false,
    })
  }

  async stop() {
    this.logger.info('Stopping Libp2pAuth service')

    // Clear the unblock interval
    clearInterval(this.unblockInterval)

    // End the outbound stream queue to prevent further pushes
    this.outboundStreamQueue.end()

    // Close all outbound streams
    for (const peerId in this.outboundStreams) {
      await this.closeOutboundStream(peerId, true)
    }

    // Close all inbound streams
    for (const peerId in this.inboundStreams) {
      await this.closeInboundStream(peerId, true)
    }

    // Clear buffered connections
    this.bufferedConnections = []

    // Clear auth connections
    this.authConnections = {}

    this.logger.info('Libp2pAuth service stopped')
  }

  private async openOutboundStream(peerId: PeerId, connection: Connection) {
    if (peerId.toString() in this.outboundStreams) {
      return
    }

    this.logger.info('Opening outbound stream for peer', peerId.toString())
    const outboundStream = await connection.newStream(this.protocol, {
      runOnLimitedConnection: false,
      negotiateFully: true,
    })
    const outboundPushable: Pushable<Uint8Array | Uint8ArrayList> = pushable()
    this.outboundStreams[peerId.toString()] = {
      stream: outboundStream,
      pushable: outboundPushable,
    }

    pipe(outboundPushable, outboundStream).catch((e: Error) =>
      this.logger.error(`Error opening outbound stream to ${peerId}`, e)
    )

    if (connection.direction === 'outbound') {
      await this.openInboundStream(peerId, connection)
    }

    this.authConnections[peerId.toString()].start()
  }

  private async openInboundStream(peerId: PeerId, connection: Connection) {
    if (peerId.toString() in this.inboundStreams) {
      return
    }

    this.logger.info('Opening new inbound stream for peer', peerId.toString())
    const inboundStream = await connection.newStream(this.protocol, {
      runOnLimitedConnection: false,
      negotiateFully: true,
    } as NewStreamOptions)

    this.handleIncomingMessages(peerId, inboundStream)
    this.inboundStreams[peerId.toString()] = inboundStream
  }

  private async onIncomingStream({ stream, connection }: IncomingStreamData) {
    const peerId = connection.remotePeer
    this.logger.info(`Handling existing incoming stream ${peerId.toString()}`)

    const oldStream = this.inboundStreams[peerId.toString()]
    if (oldStream) {
      this.logger.info(`Old inbound stream found!`)
      await this.closeInboundStream(peerId, true)
    }

    this.handleIncomingMessages(peerId, stream)

    this.inboundStreams[peerId.toString()] = stream
  }

  private handleIncomingMessages(peerId: PeerId, stream: Stream) {
    pipe(
      stream,
      source => decode(source),
      async source => {
        for await (const data of source) {
          try {
            if (!(peerId.toString() in this.authConnections)) {
              this.logger.error(`No auth connection established for ${peerId.toString()}`)
            } else {
              this.authConnections[peerId.toString()].deliver(data.subarray())
            }
          } catch (e) {
            this.logger.error(`Error while delivering message to ${peerId}`, e)
          }
        }
      }
    )
  }

  private sendMessage(peerId: PeerId, message: Uint8Array) {
    try {
      this.outboundStreams[peerId.toString()]?.pushable.push(
        // length-prefix encoded
        encode.single(message)
      )
    } catch (e) {
      this.logger.error(`Error while sending auth message over stream to ${peerId.toString()}`, e)
    }
  }

  // NOTE: This is not awaited by the registrar
  private async onPeerConnected(peerId: PeerId, connection: Connection) {
    if (this.joinStatus === JoinStatus.JOINING) {
      this.logger.warn(`Connection to ${peerId.toString()} will be buffered due to a concurrent join`)
      this.bufferedConnections.push({ peerId, connection })
      return
    }

    if (this.joinStatus === JoinStatus.PENDING) {
      this.joinStatus = JoinStatus.JOINING
    }

    this.logger.info(`Peer connected (direction = ${connection.direction})! (status = ${connection.status})`)

    // https://github.com/ChainSafe/js-libp2p-gossipsub/issues/398
    if (connection.status !== 'open') {
      this.logger.warn(`The connection with ${peerId.toString()} was not in an open state!`)
      return
    }

    const context = this.sigChainService.getActiveChain().context

    if (peerId.toString() in this.authConnections) {
      this.logger.info(
        `A connection with ${peerId.toString()} was already available, skipping connection initialization!`
      )
      return
    }

    const authConnection = new Auth.Connection({
      context,
      sendMessage: (message: Uint8Array) => {
        this.sendMessage(peerId, message)
      },
      createLogger: createLFALogger,
    } as ConnectionParams)

    authConnection.on('connected', () => {
      if (this.sigChainService.activeChainTeamName != null) {
        this.logger.debug(`Sending sync message because our chain is intialized`)
        const sigChain = this.sigChainService.getActiveChain()
        const team = sigChain.team
        const user = sigChain.localUserContext.user
        authConnection.emit('sync', { team, user })
      }
      this.emit(Libp2pEvents.AUTH_CONNECTED)
    })

    authConnection.on('disconnected', event => {
      this.logger.info(`LFA Disconnected!`, event)
      this.emit(Libp2pEvents.AUTH_DISCONNECTED)
    })

    authConnection.on('joined', payload => {
      const { team, user } = payload
      const sigChain = this.sigChainService.getActiveChain()
      this.logger.info(
        `${sigChain.localUserContext.user.userId}: Joined team ${team.teamName} (userid: ${user.userId})!`
      )
      if (sigChain.team == null && !this.joining) {
        this.joining = true
        this.logger.info(
          `${user.userId}: Creating SigChain for user with name ${user.userName} and team name ${team.teamName}`
        )
        this.logger.info(`${user.userId}: Updating auth context`)

        sigChain.context = {
          ...sigChain.context,
          team,
          user,
        } as Auth.MemberContext
        sigChain.team = team
        this.joining = false
      }
      if (this.joinStatus === JoinStatus.JOINING) {
        this.joinStatus = JoinStatus.JOINED
        this.unblockConnections(this.bufferedConnections, this.joinStatus)
      }
      this.emit(Libp2pEvents.AUTH_JOINED)
    })

    authConnection.on('change', payload => {
      this.emit(Libp2pEvents.AUTH_STATE_CHANGED, payload)
    })

    authConnection.on('updated', head => {
      this.logger.info('Received sync message, team graph updated', head)
    })

    // handle errors detected locally and reported to the peer
    authConnection.on('localError', error => {
      switch (error.type) {
        case Auth.connection.DEVICE_REMOVED:
        case Auth.connection.DEVICE_UNKNOWN:
        case Auth.connection.MEMBER_REMOVED:
        case Auth.connection.SERVER_REMOVED:
          this.emit(Libp2pEvents.AUTH_PEER_REMOVED, error, peerId)
          break
        case Auth.connection.IDENTITY_PROOF_INVALID:
        case Auth.connection.INVITATION_PROOF_INVALID:
        case Auth.connection.JOINED_WRONG_TEAM:
          this.emit(Libp2pEvents.AUTH_PEER_INVALID, error, peerId)
          break
        case Auth.connection.NEITHER_IS_MEMBER:
          this.emit(Libp2pEvents.AUTH_PEER_CANNOT_ADMIT, error)
          break
        case Auth.connection.TIMEOUT:
          this.emit(Libp2pEvents.AUTH_TIMEOUT, error)
          break
        default:
          this.emit(Libp2pEvents.AUTH_ERROR, error)
          break
      }
    })

    // handle errors sent by the peer
    authConnection.on('remoteError', error => {
      switch (error.type) {
        case Auth.connection.DEVICE_REMOVED:
        case Auth.connection.DEVICE_UNKNOWN:
        case Auth.connection.MEMBER_REMOVED:
        case Auth.connection.SERVER_REMOVED:
          this.emit(Libp2pEvents.AUTH_REMOVED, error)
          break
        case Auth.connection.IDENTITY_PROOF_INVALID:
        case Auth.connection.INVITATION_PROOF_INVALID:
        case Auth.connection.JOINED_WRONG_TEAM:
          this.emit(Libp2pEvents.AUTH_INVALID_PROOF, error)
          break
        case Auth.connection.TIMEOUT:
          this.emit(Libp2pEvents.AUTH_TIMEOUT, error)
          break
        default:
          this.emit(Libp2pEvents.AUTH_ERROR, error)
          break
      }
    })

    this.authConnections[peerId.toString()] = authConnection

    this.outboundStreamQueue.push({ peerId, connection })
  }

  private async onPeerDisconnected(peerId: PeerId) {
    this.logger.warn(`Disconnecting auth connection with peer ${peerId.toString()}`)
    await this.closeAuthConnection(peerId)
  }

  private async closeOutboundStream(peerId: PeerId | string, deleteRecord?: boolean) {
    this.logger.warn(`Closing outbound stream with ${peerId.toString()}`)
    const outboundStream = this.outboundStreams[peerId.toString()]

    if (outboundStream == null) {
      this.logger.warn(`Can't close outbound stream with ${peerId.toString()} as it doesn't exist`)
      return
    }

    await outboundStream.pushable.end()
    await outboundStream.stream.close().catch(e => {
      outboundStream.stream.abort(e)
    })

    if (deleteRecord) {
      delete this.outboundStreams[peerId.toString()]
    }
  }

  private async closeInboundStream(peerId: PeerId | string, deleteRecord?: boolean) {
    this.logger.warn(`Closing inbound stream with ${peerId.toString()}`)
    const inboundStream = this.inboundStreams[peerId.toString()]

    if (inboundStream == null) {
      this.logger.warn(`Can't close inbound stream with ${peerId.toString()} as it doesn't exist`)
      return
    }

    await inboundStream.close().catch(e => {
      inboundStream.abort(e)
    })

    if (deleteRecord) {
      delete this.inboundStreams[peerId.toString()]
    }
  }

  private async closeAuthConnection(peerId: PeerId) {
    this.logger.warn(`Closing auth connection with ${peerId.toString()}`)
    const connection = this.authConnections[peerId.toString()]

    if (connection == null) {
      this.logger.warn(`Can't close auth connection with ${peerId.toString()} as it doesn't exist`)
    } else {
      connection.stop()
      delete this.authConnections[peerId.toString()]
    }

    await this.closeOutboundStream(peerId, true)
    await this.closeInboundStream(peerId, true)
  }
}

export const libp2pAuth = (
  sigChainService: SigChainService,
  libp2pService: Libp2pService
): ((components: Libp2pAuthComponents) => Libp2pAuth) => {
  return (components: Libp2pAuthComponents) => new Libp2pAuth(sigChainService, libp2pService, components)
}
