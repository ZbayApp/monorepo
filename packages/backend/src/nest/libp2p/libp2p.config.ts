import { gossipsub } from '@chainsafe/libp2p-gossipsub'
import { identify, identifyPush } from '@libp2p/identify'
import { yamux } from '@chainsafe/libp2p-yamux'
import { tcp } from '@libp2p/tcp'
import { Libp2pOptions, ServiceFactoryMap } from 'libp2p'
import { FaultTolerance } from '@libp2p/interface-transport'
import { bootstrap } from '@libp2p/bootstrap'
import { kadDHT } from '@libp2p/kad-dht'
import { mplex } from '@libp2p/mplex'
import { ping } from '@libp2p/ping'
import { preSharedKey } from '@libp2p/pnet'
import { Libp2pAuth, Libp2pAuthComponents } from './libp2p.auth'
import { MAX_RELAY_DURATION_TIME_LIMIT, WEBSOCKET_CIPHER_SUITE } from './libp2p.const'
import { Libp2pDatastore } from './libp2p.datastore'
import { Libp2pNodeParams } from './libp2p.types'
import { keychain } from '@libp2p/keychain'
import { webSockets as webSocketsOverTor } from '../websocketOverTor'
import * as filters from '@libp2p/websockets/filters'
import { noise, pureJsCrypto } from '@chainsafe/libp2p-noise'
import * as os from 'os'
import { circuitRelayServer, circuitRelayTransport } from '@libp2p/circuit-relay-v2'
import { ServiceMap } from '@libp2p/interface'
import getPort from 'get-port'

export const generateServerLibp2pConfig = async (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth
): Promise<Libp2pOptions> => {
  const transport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 10_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: getTcpLocalAddress(params.peerId.peerId.toString(), 3000),
    targetPort: 3000,
    inboundConnectionUpgradeTimeout: 10_000,
    closeOnEnd: false,
    tagSuffix: 'tcp',
  })

  const extraServices: ServiceFactoryMap<ServiceMap> = {
    // @ts-ignore
    circuitRelay: circuitRelayServer({
      maxOutboundStopStreams: 2048,
      maxInboundHopStreams: 2048,
      reservations: {
        maxReservations: 10000,
        defaultDataLimit: BigInt(MAX_RELAY_DURATION_TIME_LIMIT),
        defaultDurationLimit: MAX_RELAY_DURATION_TIME_LIMIT,
      },
    }),
  }

  return generateLibp2pConfig(
    {
      ...params,
      transport: [transport],
      listenAddresses: [...params.listenAddresses, '/p2p-circuit'],
    },
    datastore,
    libp2pAuth,
    extraServices
  )
}

export const generatePeerLibp2pConfig = async (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth
): Promise<Libp2pOptions> => {
  const torTransport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 30_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: params.localAddress,
    targetPort: params.targetPort,
    inboundConnectionUpgradeTimeout: 30_000,
    closeOnEnd: false,
    tagSuffix: 'tor',
  })
  const wsPort = await getPort()
  const tcpTransport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 10_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: getTcpLocalAddress(params.peerId.peerId.toString(), wsPort),
    targetPort: wsPort,
    inboundConnectionUpgradeTimeout: 10_000,
    closeOnEnd: false,
    tagSuffix: 'tcp',
  })
  const relayTransport = circuitRelayTransport({
    reservationConcurrency: 1024,
  })

  return generateLibp2pConfig(
    {
      ...params,
      transport: [torTransport, tcpTransport, relayTransport],
      listenAddresses: [
        ...params.listenAddresses,
        getTcpListenAddress(params.peerId.peerId.toString(), wsPort),
        '/p2p-circuit',
      ],
    },
    datastore,
    libp2pAuth
  )
}

const generateLibp2pConfig = (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth,
  extraServices?: ServiceFactoryMap<ServiceMap>
): Libp2pOptions => {
  return {
    start: false,
    datastore: datastore.getDatastoreInstance(),
    connectionManager: {
      maxConnections: 20, // TODO: increase?
      dialTimeout: 120_000,
      maxParallelDials: 10,
      inboundUpgradeTimeout: 30_000,
      outboundUpgradeTimeout: 30_000,
      protocolNegotiationTimeout: 10_000,
      maxDialQueueLength: 500,
      reconnectRetries: 25,
    },
    privateKey: params.peerId.privKey,
    addresses: { listen: params.listenAddresses },
    connectionMonitor: {
      // ISLA: we should consider making this true if pings are reliable going forward
      abortConnectionOnPingFailure: false,
      pingInterval: 60_000,
      enabled: true,
    },
    connectionProtector:
      params.useConnectionProtector || params.useConnectionProtector == null
        ? preSharedKey({ psk: params.psk })
        : undefined,
    // @ts-ignore
    peerDiscovery: params.headless
      ? undefined
      : [
          bootstrap({
            list: ['/ip4/127.0.0.1/tcp/3000/ws/p2p/12D3KooWHPcGQbqgwdv386FmFarY27kCiYcYtxgFuBQiEQnuBqGL'],
          }),
        ],
    connectionGater: {
      // Allow private addresses for local testing
      denyDialMultiaddr: async () => false,
    },
    streamMuxers: [
      // mplex({
      //   disconnectThreshold: 20,
      //   maxInboundStreams: 1024,
      //   maxOutboundStreams: 1024,
      //   maxStreamBufferSize: 26214400,
      //   maxUnprocessedMessageQueueSize: 104857600,
      //   maxMsgSize: 10485760,
      //   // @ts-expect-error This is part of the config interface but it isn't typed that way
      //   closeTimeout: 15_000,
      // }),
      yamux({
        maxInboundStreams: 1024,
        maxOutboundStreams: 1024,
        maxMessageSize: 10485760,
      }),
    ],
    // @ts-ignore
    connectionEncrypters: [noise({ crypto: pureJsCrypto, staticNoiseKey: params.peerId.noiseKey })],
    transports: params.transport!,
    transportManager: {
      faultTolerance: FaultTolerance.NO_FATAL,
    },
    services: {
      ...(extraServices ?? {}),
      auth: libp2pAuth,
      ping: ping({ timeout: 30_000 }),
      pubsub: gossipsub({
        // neccessary to run a single peer
        allowPublishToZeroTopicPeers: true,
        fallbackToFloodsub: false,
        emitSelf: true,
        debugName: params.peerId.peerId.toString(),
        doPX: true,
      }),
      identify: identify({ timeout: 30_000, maxInboundStreams: 128, maxOutboundStreams: 128 }),
      identifyPush: identifyPush({ timeout: 30_000, maxInboundStreams: 128, maxOutboundStreams: 128 }),
      keychain: keychain(),
      dht: kadDHT({
        allowQueryWithZeroPeers: false,
        clientMode: false,
        initialQuerySelfInterval: 500,
        providers: {
          cacheSize: 1024,
        },
        maxInboundStreams: 128,
        maxOutboundStreams: 128,
      }),
    },
  }
}

export type IPMap = {
  [interfaceName: string]: string[]
}

export const DEFAULT_NETWORK_INTERFACE = 'en0'

/*
Shamelessly copied from https://stackoverflow.com/a/8440736
*/
export const getIpAddresses = (): IPMap => {
  const interfaces = os.networkInterfaces()
  const results: IPMap = {} // Or just '{}', an empty object

  for (const name in interfaces) {
    for (const net of interfaces[name]!) {
      // Skip over non-IPv4 and internal (i.e. 127.0.0.1) addresses
      // 'IPv4' is in Node <= 17, from 18 it's a number 4 or 6
      const familyV4Value = typeof net.family === 'string' ? 'IPv4' : 4
      if (net.family === familyV4Value && !net.internal) {
        if (!results[name]) {
          results[name] = []
        }
        results[name].push(net.address)
      }
    }
  }

  return results
}

export const getTcpListenAddress = (peerId: string, port: number): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/${port}/ws/p2p/${peerId}`
}

export const getTcpLocalAddress = (peerId: string, port: number): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/${port}/ws/p2p/${peerId}`
}

export const getRelayPeerListenAddress = (peerId: string): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/3000/ws/p2p/12D3KooWHPcGQbqgwdv386FmFarY27kCiYcYtxgFuBQiEQnuBqGL/p2p-circuit/p2p/${peerId}`
}
