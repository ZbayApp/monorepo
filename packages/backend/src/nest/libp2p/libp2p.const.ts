export const BITSWAP_PROTOCOL = '/quiet/ipfs/bitswap/1.2.0'
export const WEBSOCKET_CIPHER_SUITE = 'TLS_AES_256_GCM_SHA384'
