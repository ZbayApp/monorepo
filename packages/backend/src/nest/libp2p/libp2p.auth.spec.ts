import { Test, TestingModule } from '@nestjs/testing'
import { TestModule } from '../common/test.module'
import { getLocalLibp2pInstanceParams } from '../common/utils'
import { Libp2pModule } from './libp2p.module'
import { SigChainModule } from '../auth/sigchain.service.module'
import { Libp2pService } from './libp2p.service'
import { SigChainService } from '../auth/sigchain.service'
import { Libp2pEvents, Libp2pNodeParams } from './libp2p.types'
import { createLogger } from '../common/logger'
import { Libp2pAuth } from './libp2p.auth'

const logger = createLogger('libp2p:libp2p.auth.spec')

const attachEventListeners = (libp2pService: Libp2pService, timeline: string[], instanceName: string) => {
  // loop over all enum Libp2pEvents and attach event listeners
  for (const event of Object.values(Libp2pEvents)) {
    libp2pService.on(event, () => {
      timeline.push(`${instanceName}: ${event}`)
    })
  }
}

describe('Libp2pAuth', () => {
  const teamName: string = 'team'
  const userA: string = 'instanceA'
  const userB: string = 'instanceB'
  const eventTimeline: string[] = []
  const eventTimelineA: string[] = []
  const eventTimelineB: string[] = []
  let moduleA: TestingModule
  let moduleB: TestingModule
  let libp2pServiceA: Libp2pService
  let libp2pServiceB: Libp2pService
  let sigchainServiceA: SigChainService
  let sigchainServiceB: SigChainService
  let paramsA: Libp2pNodeParams
  let paramsB: Libp2pNodeParams

  beforeAll(async () => {
    moduleA = await Test.createTestingModule({
      imports: [TestModule, Libp2pModule, SigChainModule],
    }).compile()

    moduleB = await Test.createTestingModule({
      imports: [TestModule, Libp2pModule, SigChainModule],
    }).compile()
    sigchainServiceA = await moduleA.resolve(SigChainService)
    sigchainServiceB = await moduleB.resolve(SigChainService)
    libp2pServiceA = await moduleA.resolve(Libp2pService)
    libp2pServiceB = await moduleB.resolve(Libp2pService)

    attachEventListeners(libp2pServiceA, eventTimeline, 'A')
    attachEventListeners(libp2pServiceB, eventTimeline, 'B')
    attachEventListeners(libp2pServiceA, eventTimelineA, 'A')
    attachEventListeners(libp2pServiceB, eventTimelineB, 'B')

    const singlePSK = Libp2pService.generateLibp2pPSK().fullKey

    paramsA = {
      ...(await getLocalLibp2pInstanceParams()),
      psk: singlePSK,
      instanceName: userA,
    }

    paramsB = {
      ...(await getLocalLibp2pInstanceParams()),
      psk: singlePSK,
      instanceName: userB,
    }

    // Create chain for instance A
    await sigchainServiceA.createChain(teamName, userA, true)
    expect(sigchainServiceA.activeChainTeamName).toBe(teamName)

    // Create invitation from A -> B
    const inviteResult = await sigchainServiceA.getActiveChain().invites.createLongLivedUserInvite()
    await sigchainServiceB.createChainFromInvite(userB, teamName, inviteResult.seed, true)
    expect(sigchainServiceB.activeChainTeamName).toBe(teamName)
  })

  afterAll(async () => {
    await libp2pServiceA.libp2pInstance?.stop()
    await libp2pServiceB.libp2pInstance?.stop()
    await moduleA.close()
    await moduleB.close()
  })

  it('create two instances of libp2p', async () => {
    logger.info('Creating libp2p instances')

    logger.info('paramsA', paramsA)
    await libp2pServiceA.createInstance(paramsA)
    expect(libp2pServiceA.libp2pInstance).not.toBeNull()
    expect(libp2pServiceA?.libp2pInstance?.peerId.toString()).toBe(paramsA.peerId.peerId.toString())

    logger.info('paramsB', paramsB)
    expect(paramsA.psk).toEqual(paramsB.psk)
    await libp2pServiceB.createInstance(paramsB)
    expect(libp2pServiceB.libp2pInstance).not.toBeNull()
    expect(libp2pServiceB?.libp2pInstance?.peerId.toString()).toBe(paramsB.peerId.peerId.toString())
  })

  it(
    'dials each other',
    async () => {
      // Wait for chain init event on B
      const waitForChainInitializedB = new Promise<void>(resolve => {
        libp2pServiceB.once(Libp2pEvents.AUTH_JOINED, () => {
          logger.info('libp2pServiceB initialized chain')
          resolve()
        })
      })
      await libp2pServiceB.dialPeer(paramsA.localAddress)
      await waitForChainInitializedB

      expect(eventTimeline).toMatchSnapshot('event timeline after dialing')
      expect(eventTimelineA).toMatchSnapshot('instanceA event timeline after join')
      expect(eventTimelineB).toMatchSnapshot('instanceB event timeline after join')
    },
    1000 * 60 * 5
  )
  it('gracefully disconnects', async () => {
    const disconnectEventsA: string[] = []
    const disconnectEventsB: string[] = []
    attachEventListeners(libp2pServiceA, disconnectEventsA, 'A')
    attachEventListeners(libp2pServiceB, disconnectEventsB, 'B')

    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)

      libp2pServiceA.once(Libp2pEvents.PEER_DISCONNECTED, () => {
        logger.info('test listener heard libp2pServiceA disconnected')
        clearTimeout(timeout)
        resolve()
      })
      libp2pServiceB.once(Libp2pEvents.PEER_DISCONNECTED, () => {
        logger.info('test listener heard libp2pServiceB disconnected')
        clearTimeout(timeout)
        resolve()
      })
      libp2pServiceB.libp2pInstance?.stop()
    })

    expect(disconnectEventsA).toMatchSnapshot('disconnectEventsA')
    expect(disconnectEventsB).toMatchSnapshot('disconnectEventsB')
    expect(eventTimeline).toMatchSnapshot('eventTimeline after disconnection')
  })
  // it(
  //   'removes user from chain and disconnects',
  //   async () => {
  //     const removeEventsA: string[] = []
  //     const removeEventsB: string[] = []
  //     attachEventListeners(libp2pServiceA, removeEventsA, 'A')
  //     attachEventListeners(libp2pServiceB, removeEventsB, 'B')

  //     await new Promise<void>((resolve, reject) => {
  //       const timeout = setTimeout(() => {
  //         reject(new Error('PEER_DISCONNECTED or chain removal events did not occur within expected time.'))
  //       }, 70_000)

  //       libp2pServiceB.once(Libp2pEvents.PEER_DISCONNECTED, () => {
  //         logger.info('test listener heard libp2pServiceB disconnected')
  //         clearTimeout(timeout)
  //         resolve()
  //       })
  //       sigchainServiceA.getActiveChain().team?.remove(sigchainServiceB.getActiveChain().localUserContext.user.userId)
  //     })

  //     expect(removeEventsA).toMatchSnapshot('removeEventsA')
  //     expect(removeEventsB).toMatchSnapshot('removeEventsB')
  //     expect(eventTimeline).toMatchSnapshot('eventTimeline after removal')
  //   },
  //   1000 * 60 * 5
  // )
})
