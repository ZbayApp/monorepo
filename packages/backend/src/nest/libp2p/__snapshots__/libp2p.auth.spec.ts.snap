// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Libp2pAuth dials each other: event timeline after dialing 1`] = `
[
  "A: peerConnected",
  "B: peerConnected",
  "A: authStateChanged",
  "B: authStateChanged",
  "B: authStateChanged",
  "A: authStateChanged",
  "A: authStateChanged",
  "B: authStateChanged",
  "B: authJoined",
  "B: authStateChanged",
]
`;

exports[`Libp2pAuth dials each other: instanceA event timeline after join 1`] = `
[
  "A: peerConnected",
  "A: authStateChanged",
  "A: authStateChanged",
  "A: authStateChanged",
]
`;

exports[`Libp2pAuth dials each other: instanceB event timeline after join 1`] = `
[
  "B: peerConnected",
  "B: authStateChanged",
  "B: authStateChanged",
  "B: authStateChanged",
  "B: authJoined",
  "B: authStateChanged",
]
`;

exports[`Libp2pAuth gracefully disconnects: disconnectEventsA 1`] = `
[
  "A: authStateChanged",
  "A: peerDisconnected",
]
`;

exports[`Libp2pAuth gracefully disconnects: disconnectEventsB 1`] = `[]`;

exports[`Libp2pAuth gracefully disconnects: eventTimeline after disconnection 1`] = `
[
  "A: peerConnected",
  "B: peerConnected",
  "A: authStateChanged",
  "B: authStateChanged",
  "B: authStateChanged",
  "A: authStateChanged",
  "A: authStateChanged",
  "B: authStateChanged",
  "B: authJoined",
  "B: authStateChanged",
  "A: authStateChanged",
  "A: peerDisconnected",
]
`;
